gcpFS/FileOps.go

```diff
@@ -10,58 +10,47 @@ import (
 	"time"
 
 	"cloud.google.com/go/storage"
-	"github.com/ninjamarcus/ninjaStorage/models"
 	"google.golang.org/api/iterator"
+
+	_ "github.com/ninjamarcus/ninjaStorage/Interfaces" /*ninjaStorage*/
+	"github.com/ninjamarcus/ninjaStorage/models"
 )
 
 type GCPFS struct {
-	//storage is the gcp storage client
+	//storage is the GCP storage client
 	client *storage.Client
 	config *models.GCPFSConfig
 	ctx    context.Context
 }
 
-type GCPControls interface {
-	NewGCPStorage(fs *models.GCPFSConfig) (*GCPFS, error)
-	Delete(g *GCPFS, filePath string) error
-	Move(g *GCPFS, filePathFrom string, filePathTo string) error
-	Copy(g *GCPFS, filePathFrom string, filePathTo string) error
-	Find()
-	Write(g *GCPFS, data []byte, filePath string, metaData *models.FileMetaData) (*models.FileMetaData, error)
-	List(g *GCPFS, prefix string) (map[string]*models.FileMetaData, error)
-	Read(g *GCPFS, filePath string) ([]byte, *models.FileMetaData, error)
-}
-
-type GCPController struct{}
-
 // NewGCPStorage TO Connect successfully you need to have exported your service account.json file
 // as the environment variable GOOGLE_APPLICATION_CREDENTIALS
-func (gcp *GCPController) NewGCPStorage(fs *models.GCPFSConfig) (*GCPFS, error) {
-	if err := fs.Validate(); err != nil {
+func NewGCPStorage(config *models.GCPFSConfig) (*GCPFS, error) {
+	if err := config.Validate(); err != nil {
 		return &GCPFS{}, err
 	}
-	gcpfs := &GCPFS{config: fs}
-	if err := gcpfs.connectToGCPStorage(); err != nil {
+
+	g := &GCPFS{config: config}
+	if err := g.Connect(); err != nil {
 		return &GCPFS{}, err
 	}
-
-	return gcpfs, nil
+	return g, nil
 }
 
-// Connect to the client
-func (g *GCPFS) connectToGCPStorage() error {
+// Connect to the GCP client
+func (g *GCPFS) Connect() error {
 	ctx := context.Background()
 	client, err := storage.NewClient(ctx)
 	if err != nil {
 		return err
 	}
-	g.client = client
 	g.ctx = ctx
-	defer client.Close()
+	g.client = client
+	client.Close()
```

Comment:
Should this not be a defer client.Close() as opposed to a close, as won't this just shutdown the connection? Im happy to be wrong here. The examples in googles docs seem to suggest that is the way to go - https://cloud.google.com/storage/docs/creating-buckets#permissions-client-libraries

Reply:
a deferred call is "deferred" until immediately before the enclosing function returns. As this is already the last statement in said function, there is no point in deferring it.

Reply:
I suspect that the client.Close() is actually redundant here. It looks like the client will (re)establish the connection when it is actually used. Also, according to the docs, it is not "necessary" to call close prior to the program exiting.